src/glob.zig

```diff
@@ -0,0 +1,3032 @@
+        var patternComponents = ArrayList(Component){};
+        try GlobWalker.buildPatternComponents(
+            arena,
+            &patternComponents,
+            pattern,
+            &this.cp_len,
+            &this.pattern_codepoints,
+            &this.has_relative_components,
+        );
+
+        this.patternComponents = patternComponents;
+        this.pattern = pattern;
+        this.arena = arena.*;
+        this.cwd = cwd;
+        this.dot = dot;
+        this.absolute = absolute;
+        this.follow_symlinks = follow_symlinks;
+        this.error_on_broken_symlinks = error_on_broken_symlinks;
+        this.only_files = only_files;
+    }
+
+    pub fn deinit(this: *GlobWalker) void {
+        this.arena.deinit();
+    }
+
+    pub fn handleSysErrWithPath(
+        this: *GlobWalker,
+        err: Syscall.Error,
+        path_buf: [:0]const u8,
+    ) Syscall.Error {
+        @memcpy(this.pathBuf[0 .. path_buf.len + 1], @as([]const u8, @ptrCast(path_buf[0 .. path_buf.len + 1])));
+        return err.withPath(this.pathBuf[0 .. path_buf.len + 1]);
+    }
+
+    pub fn walk(this: *GlobWalker) !Maybe(u0) {
+        if (this.patternComponents.items.len == 0) return .{ .result = 0 };
+        var path_buf: [bun.MAX_PATH_BYTES]u8 = std.mem.zeroes([bun.MAX_PATH_BYTES]u8);
+
+        const root_path = this.cwd;
+        @memcpy(path_buf[0..root_path.len], root_path[0..root_path.len]);
+        path_buf[root_path.len] = 0;
+        this.root_cwd_slice = root_path;
+        var cwd_fd = switch (Syscall.open(@ptrCast(path_buf[0 .. root_path.len + 1]), std.os.O.DIRECTORY | std.os.O.RDONLY, 0)) {
+            .err => |err| return .{ .err = this.handleSysErrWithPath(err, @ptrCast(path_buf[0 .. root_path.len + 1])) },
+            .result => |fd| fd,
+        };
+        defer {
+            _ = Syscall.closeAllowingStdoutAndStderr(cwd_fd);
+        }
+        // Handle root first
+        {
+            const root_work_item = WorkItem.new(root_path, 0, .directory);
+            switch (try this.handleDirEntriesRoot(&root_work_item, &path_buf, cwd_fd)) {
+                .err => |err| return .{ .err = err },
+                else => {},
+            }
+        }
+
+        while (this.workbuf.items.len > 0) {
+            const work_item: WorkItem = this.workbuf.pop();
+            switch (work_item.kind) {
+                .directory => switch (try this.handleDirEntries(&work_item, &path_buf, cwd_fd)) {
+                    .err => |err| return .{ .err = err },
+                    else => {},
+                },
+                .symlink => switch (try this.handleSymlink(&work_item, &path_buf)) {
+                    .err => |err| return .{ .err = err },
+                    else => {},
+                },
+            }
+        }
+
+        return .{ .result = 0 };
+    }
+
+    pub fn handleSymlink(
+        this: *GlobWalker,
+        work_item: *const WorkItem,
+        scratch_path_buf: *[bun.MAX_PATH_BYTES]u8,
+    ) !Maybe(u0) {
```

Comment:
why is this a `Maybe(u0)` rather than a `Maybe(void)`? Usually `u0` is useful for things which we might want to think of like a number but disable it